File: dashboard/routes/machines.py
"""Adding, inspecting, driving and removing machines."""

import time

from flask import Blueprint, g, jsonify, render_template, request

import fleet
import guards
from db import connect
from machines import (budgeted_machine_count, computer_views, create_one_computer,
                      expand_names, get_computer, list_computers, touch_active,
                      wake_and_wait)
from security import require_token
from settings import MAX_CLIPBOARD_KB, SHOT_TTL
from tasks import active_task_by_computer

bp = Blueprint("machines", __name__)


@bp.route("/partials/fleet")
def partial_fleet():
    busy = active_task_by_computer()
    computers = computer_views(list_computers())
    for view in computers:
        view["active_task"] = busy.get(view["name"])
    return render_template("_fleet.html", computers=computers,
                           shot_token=int(time.time() // SHOT_TTL))


@bp.route("/api/v1/computers", methods=["GET"])
def api_computers_list():
    return jsonify({"ok": True, "data": computer_views(list_computers()), "error": None})


@bp.route("/api/v1/computers", methods=["POST"])
@require_token
def api_computers_create():
    payload = request.get_json(silent=True) or {}
    name = (payload.get("name") or "").strip()
    snapshot_name = (payload.get("snapshot") or "").strip()
    if not name:
        return jsonify({"ok": False, "data": None, "error": "name is required"}), 400

    image = None
    if snapshot_name:
        conn = connect()
        snap = conn.execute("SELECT * FROM snapshots WHERE name = ?", (snapshot_name,)).fetchone()
        conn.close()
        if not snap:
            return jsonify({"ok": False, "data": None,
                            "error": f"unknown snapshot '{snapshot_name}'"}), 400
        image = snap["image"]

    try:
        names = expand_names(name)
    except ValueError as exc:
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 400

    fleet_size = budgeted_machine_count()
    for ok, msg in (guards.check_memory(len(names), fleet_size), guards.check_disk()):
        if not ok:
            return jsonify({"ok": False, "data": None, "error": msg}), 507

    conn = connect()
    created, errors = [], []
    for n in names:
        try:
            created.append(create_one_computer(conn, n, image))
        except ValueError as exc:
            errors.append({"name": n, "error": str(exc)})
        except Exception as exc:  # noqa: BLE001
            errors.append({"name": n, "error": f"failed to start containers: {exc}"})
    conn.close()

    if not created:
        msg = errors[0]["error"] if errors else "nothing created"
        code = 409 if errors and "already exists" in errors[0]["error"] else 500
        return jsonify({"ok": False, "data": {"errors": errors}, "error": msg}), code

    # Creation is the one mutation with no machine to look up beforehand, so
    # the audit target has to be named here rather than by get_computer.
    g.audit_target = names[0] if len(names) == 1 else None
    g.audit_detail = (f"created {len(created)}" +
                      (f", {len(errors)} failed" if errors else "") +
                      (f" ({', '.join(c['name'] for c in created[:8])})"
                       if len(names) > 1 else ""))

    # Single-name requests keep returning the bare object they always did.
    data = created[0] if len(names) == 1 else {"created": created, "errors": errors}
    return jsonify({"ok": True, "data": data, "error": None}), 201


@bp.route("/api/v1/computers/<int:comp_id>", methods=["PATCH"])
@require_token
def api_computers_rename(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    payload = request.get_json(silent=True) or {}

    # Same endpoint also flips the boolean flags, which carry no name.
    for field in ("reserved", "no_suspend"):
        if field in payload and "name" not in payload:
            flag = 1 if str(payload[field]).lower() in ("1", "true", "yes") else 0
            conn = connect()
            conn.execute(f"UPDATE computers SET {field} = ? WHERE id = ?", (flag, comp_id))
            conn.commit()
            conn.close()
            return jsonify({"ok": True, "data": {"id": comp_id, field: bool(flag)},
                            "error": None})

    new_name = (payload.get("name") or "").strip()
    if not new_name:
        return jsonify({"ok": False, "data": None, "error": "name is required"}), 400

    conn = connect()
    clash = conn.execute("SELECT 1 FROM computers WHERE name = ? AND id != ?", (new_name, comp_id)).fetchone()
    if clash:
        conn.close()
        return jsonify({"ok": False, "data": None, "error": f"'{new_name}' already exists"}), 409
    # Only the display name changes — the slug stays, so containers and any
    # in-flight task keep pointing at the same machine.
    conn.execute("UPDATE computers SET name = ? WHERE id = ?", (new_name, comp_id))
    conn.execute("UPDATE tasks SET desktop = ? WHERE desktop = ?", (new_name, comp["name"]))
    conn.commit()
    conn.close()
    return jsonify({"ok": True, "data": {"id": comp_id, "name": new_name}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/restart", methods=["POST"])
@require_token
def api_computer_restart(comp_id: int):
    """Recreate a machine's containers in place.

    A machine can end up with its containers gone or wedged — the host
    rebooted, someone ran `docker rm`, the bridge died. Before this the only
    cure was delete-and-recreate, which lost the machine's name and port.
    """
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    try:
        # keep_home: a restart is meant to fix the machine, not wipe your work.
        fleet.destroy_computer(comp["slug"], keep_home=True)
        fleet.create_computer(comp["slug"], comp["novnc_port"],
                              comp["vnc_password"], image=comp["image"])
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None,
                        "error": f"failed to restart: {exc}"}), 500
    return jsonify({"ok": True, "data": {"id": comp_id, "restarted": True}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/sleep", methods=["POST"])
@require_token
def api_computer_sleep(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    conn = connect()
    busy = conn.execute(
        "SELECT COUNT(*) AS n FROM tasks WHERE desktop = ? AND status IN ('PENDING','RUNNING')",
        (comp["name"],),
    ).fetchone()["n"]
    conn.close()
    if busy:
        return jsonify({"ok": False, "data": None,
                        "error": f"{comp['name']} has {busy} task(s) still running"}), 409
    try:
        fleet.suspend_computer(comp["slug"])
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": f"failed to sleep: {exc}"}), 500
    return jsonify({"ok": True, "data": {"id": comp_id, "sleeping": True}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/wake", methods=["POST"])
@require_token
def api_computer_wake(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    try:
        result = wake_and_wait(comp)
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": f"failed to wake: {exc}"}), 500
    # Started but not yet answering is not an error — the screen will come up
    # a moment later — so this reports readiness rather than failing.
    return jsonify({"ok": True, "data": {"id": comp_id, "sleeping": False, **result},
                    "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/clipboard", methods=["GET"])
def api_clipboard_get(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    try:
        text = fleet.get_clipboard(comp["slug"])
    except fleet.ClipboardUnavailable as exc:
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 503
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 500
    return jsonify({"ok": True, "data": {"text": text}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/clipboard", methods=["POST"])
@require_token
def api_clipboard_set(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    payload = request.get_json(silent=True) or {}
    text = payload.get("text")
    if text is None:
        return jsonify({"ok": False, "data": None, "error": "text is required"}), 400
    text = str(text)
    if len(text.encode("utf-8")) > MAX_CLIPBOARD_KB * 1024:
        return jsonify({"ok": False, "data": None,
                        "error": f"clipboard text over {MAX_CLIPBOARD_KB} KB"}), 413
    # "paste" also presses Ctrl+V, which is what makes Arabic typing work at
    # all — see fleet.paste_text.
    press = str(payload.get("paste", "")).lower() in ("1", "true", "yes")
    # Size and intent, not the text — an audit trail that archives everything
    # anyone pasted is its own kind of problem.
    g.audit_detail = f"{len(text.encode('utf-8'))} bytes, paste={press}"
    try:
        (fleet.paste_text if press else fleet.set_clipboard)(comp["slug"], text)
    except fleet.ClipboardUnavailable as exc:
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 503
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 500
    touch_active(comp_id)
    return jsonify({"ok": True, "data": {"id": comp_id, "bytes": len(text.encode("utf-8")),
                                         "pasted": press}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>", methods=["DELETE"])
@require_token
def api_computers_delete(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    try:
        fleet.destroy_computer(comp["slug"])
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": f"failed to remove containers: {exc}"}), 500
    conn = connect()
    conn.execute("DELETE FROM computers WHERE id = ?", (comp_id,))
    conn.commit()
    conn.close()
    return jsonify({"ok": True, "data": {"id": comp_id, "removed": True}, "error": None})


@bp.route("/api/v1/computers/<int:comp_id>/inventory", methods=["GET"])
def api_computers_inventory(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    try:
        data = fleet.get_inventory(comp["slug"])
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 500
    return jsonify({"ok": True, "data": data, "error": None})


@bp.route("/partials/computers/<int:comp_id>/inventory")
def partial_inventory(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return "<div class='text-red-400 text-xs'>computer not found</div>", 404
    try:
        inv = fleet.get_inventory(comp["slug"])
    except Exception as exc:  # noqa: BLE001
        inv = {"error": str(exc)}
    return render_template("_inventory.html", comp=comp, inv=inv,
                           vnc_password=comp["vnc_password"])


@bp.route("/api/v1/computers/<int:comp_id>/exec", methods=["POST"])
@require_token
def api_computers_exec(comp_id: int):
    comp = get_computer(comp_id)
    if not comp:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    payload = request.get_json(silent=True) or {}
    command = (payload.get("command") or "").strip()
    if not command:
        return jsonify({"ok": False, "data": None, "error": "command is required"}), 400
    g.audit_detail = command[:500]
    try:
        result = fleet.exec_in_desktop(comp["slug"], command)
    except Exception as exc:  # noqa: BLE001
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 500
    return jsonify({"ok": True, "data": result, "error": None})


@bp.route("/api/v1/fleet")
def api_fleet():
    return jsonify({"ok": True, "data": computer_views(list_computers()), "error": None})

File: dashboard-ts/src/routes/machines.ts
/** Adding, inspecting, driving and removing machines. */

import { Hono, type Context } from "hono";
import { providerFor, ClipboardUnavailable } from "./../providers";
import * as guards from "./../guards";
import { one, run } from "./../db";
import {
  budgetedMachineCount,
  computerViews,
  createOneComputer,
  expandNames,
  getComputer,
  listComputers,
  touchActive,
  wakeAndWait,
  ValidationError,
  type Computer,
} from "./../machines";
import { activeTaskByComputer } from "./../tasks";
import { render } from "./../templates";
import { requireToken } from "./../security";
import { MAX_CLIPBOARD_KB, SHOT_TTL } from "./../settings";
import {
  browserHost,
  fail,
  intParam,
  jsonBody,
  notFound,
  ok,
  type Env,
} from "./../http";

export const machines = new Hono<Env>();

/**
 * Nearly every per-machine handler starts here, so naming the machine for the
 * audit log once means no handler has to remember to do it.
 */
function loadComputer(c: Context<Env>): Computer | null {
  const id = intParam(c, "id");
  if (id === null) return null;
  const comp = getComputer(id);
  if (comp) c.set("auditTarget", comp.name);
  return comp;
}

const truthy = (v: unknown) => ["1", "true", "yes"].includes(String(v).toLowerCase());

machines.get("/partials/fleet", async (c) => {
  const busy = activeTaskByComputer();
  const computers = await computerViews(listComputers(), browserHost(c));
  for (const view of computers) view.active_task = busy[view.name];
  return c.html(
    render("_fleet.html", {
      computers,
      shot_token: Math.floor(Date.now() / 1000 / SHOT_TTL),
    }),
  );
});

machines.get("/api/v1/computers", async (c) =>
  ok(c, await computerViews(listComputers(), browserHost(c))),
);

machines.get("/api/v1/fleet", async (c) =>
  ok(c, await computerViews(listComputers(), browserHost(c))),
);

machines.post("/api/v1/computers", requireToken, async (c) => {
  const payload = await jsonBody(c);
  const name = String(payload.name ?? "").trim();
  const snapshotName = String(payload.snapshot ?? "").trim();
  if (!name) return fail(c, "name is required");

  let image: string | null = null;
  if (snapshotName) {
    const snap = one<{ image: string }>(
      "SELECT * FROM snapshots WHERE name = ?",
      snapshotName,
    );
    if (!snap) return fail(c, `unknown snapshot '${snapshotName}'`);
    image = snap.image;
  }

  let names: string[];
  try {
    names = expandNames(name);
  } catch (err: any) {
    return fail(c, String(err.message));
  }

  const fleetSize = await budgetedMachineCount();
  for (const [passed, msg] of [
    guards.checkMemory(names.length, fleetSize),
    guards.checkDisk(),
  ]) {
    if (!passed) return fail(c, msg, 507);
  }

  const created: unknown[] = [];
  const errors: { name: string; error: string }[] = [];
  for (const n of names) {
    try {
      created.push(await createOneComputer(n, image));
    } catch (err: any) {
      errors.push({
        name: n,
        error:
          err instanceof ValidationError
            ? String(err.message)
            : `failed to start containers: ${err?.message ?? err}`,
      });
    }
  }

  if (!created.length) {
    const msg = errors[0]?.error ?? "nothing created";
    const code = msg.includes("already exists") ? 409 : 500;
    return fail(c, msg, code, { errors });
  }

  // Creation is the one mutation with no machine to look up beforehand, so the
  // audit target has to be named here rather than by loadComputer.
  c.set("auditTarget", names.length === 1 ? names[0] : null);
  c.set(
    "auditDetail",
    `created ${created.length}` +
      (errors.length ? `, ${errors.length} failed` : "") +
      (names.length > 1
        ? ` (${created.slice(0, 8).map((x: any) => x.name).join(", ")})`
        : ""),
  );

  // Single-name requests keep returning the bare object they always did.
  const data = names.length === 1 ? created[0] : { created, errors };
  return ok(c, data, 201);
});

machines.patch("/api/v1/computers/:id", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const payload = await jsonBody(c);

  // Same endpoint also flips the boolean flags, which carry no name.
  for (const field of ["reserved", "no_suspend"] as const) {
    if (field in payload && !("name" in payload)) {
      const flag = truthy(payload[field]) ? 1 : 0;
      run(`UPDATE computers SET ${field} = ? WHERE id = ?`, flag, comp.id);
      return ok(c, { id: comp.id, [field]: !!flag });
    }
  }

  const newName = String(payload.name ?? "").trim();
  if (!newName) return fail(c, "name is required");
  const clash = one(
    "SELECT 1 AS x FROM computers WHERE name = ? AND id != ?",
    newName,
    comp.id,
  );
  if (clash) return fail(c, `'${newName}' already exists`, 409);

  // Only the display name changes — the slug stays, so containers and any
  // in-flight task keep pointing at the same machine.
  run("UPDATE computers SET name = ? WHERE id = ?", newName, comp.id);
  run("UPDATE tasks SET desktop = ? WHERE desktop = ?", newName, comp.name);
  return ok(c, { id: comp.id, name: newName });
});

/**
 * Recreate a machine's containers in place.
 *
 * A machine can end up with its containers gone or wedged — the host rebooted,
 * someone ran `docker rm`, the bridge died. Before this the only cure was
 * delete-and-recreate, which lost the machine's name and port.
 */
machines.post("/api/v1/computers/:id/restart", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  try {
    // keep_home: a restart is meant to fix the machine, not wipe your work.
    const backend = providerFor(comp);
    await backend.destroyComputer(comp.slug, true);
    await backend.createComputer(comp.slug, comp.novnc_port, comp.vnc_password, comp.image);
  } catch (err: any) {
    return fail(c, `failed to restart: ${err?.message ?? err}`, 500);
  }
  return ok(c, { id: comp.id, restarted: true });
});

machines.post("/api/v1/computers/:id/sleep", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const busy =
    one<{ n: number }>(
      "SELECT COUNT(*) AS n FROM tasks WHERE desktop = ? AND status IN ('PENDING','RUNNING')",
      comp.name,
    )?.n ?? 0;
  if (busy) {
    return fail(c, `${comp.name} has ${busy} task(s) still running`, 409);
  }
  try {
    await providerFor(comp).suspendComputer(comp.slug);
  } catch (err: any) {
    return fail(c, `failed to sleep: ${err?.message ?? err}`, 500);
  }
  return ok(c, { id: comp.id, sleeping: true });
});

machines.post("/api/v1/computers/:id/wake", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  let result;
  try {
    result = await wakeAndWait(comp);
  } catch (err: any) {
    return fail(c, `failed to wake: ${err?.message ?? err}`, 500);
  }
  // Started but not yet answering is not an error — the screen will come up a
  // moment later — so this reports readiness rather than failing.
  return ok(c, { id: comp.id, sleeping: false, ...result });
});

machines.get("/api/v1/computers/:id/clipboard", async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  try {
    return ok(c, { text: await providerFor(comp).getClipboard(comp.slug) });
  } catch (err: any) {
    const code = err instanceof ClipboardUnavailable ? 503 : 500;
    return fail(c, String(err?.message ?? err), code);
  }
});

machines.post("/api/v1/computers/:id/clipboard", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const payload = await jsonBody(c);
  if (payload.text === undefined || payload.text === null) {
    return fail(c, "text is required");
  }
  const text = String(payload.text);
  const bytes = Buffer.byteLength(text, "utf8");
  if (bytes > MAX_CLIPBOARD_KB * 1024) {
    return fail(c, `clipboard text over ${MAX_CLIPBOARD_KB} KB`, 413);
  }
  // "paste" also presses Ctrl+V, which is what makes Arabic typing work at
  // all — see the provider's pasteText.
  const press = truthy(payload.paste ?? "");
  // Size and intent, not the text — an audit trail that archives everything
  // anyone pasted is its own kind of problem.
  c.set("auditDetail", `${bytes} bytes, paste=${press}`);
  try {
    const backend = providerFor(comp);
    if (press) await backend.pasteText(comp.slug, text);
    else await backend.setClipboard(comp.slug, text);
  } catch (err: any) {
    const code = err instanceof ClipboardUnavailable ? 503 : 500;
    return fail(c, String(err?.message ?? err), code);
  }
  touchActive(comp.id);
  return ok(c, { id: comp.id, bytes, pasted: press });
});

machines.delete("/api/v1/computers/:id", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  try {
    await providerFor(comp).destroyComputer(comp.slug);
  } catch (err: any) {
    return fail(c, `failed to remove containers: ${err?.message ?? err}`, 500);
  }
  run("DELETE FROM computers WHERE id = ?", comp.id);
  return ok(c, { id: comp.id, removed: true });
});

machines.get("/api/v1/computers/:id/inventory", async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  try {
    return ok(c, await providerFor(comp).getInventory(comp.slug));
  } catch (err: any) {
    return fail(c, String(err?.message ?? err), 500);
  }
});

machines.get("/partials/computers/:id/inventory", async (c) => {
  const comp = loadComputer(c);
  if (!comp) {
    return c.html("<div class='text-red-400 text-xs'>computer not found</div>", 404);
  }
  let inv: unknown;
  try {
    inv = await providerFor(comp).getInventory(comp.slug);
  } catch (err: any) {
    inv = { error: String(err?.message ?? err) };
  }
  return c.html(
    render("_inventory.html", { comp, inv, vnc_password: comp.vnc_password }),
  );
});

machines.post("/api/v1/computers/:id/exec", requireToken, async (c) => {
  const comp = loadComputer(c);
  if (!comp) return notFound(c);
  const payload = await jsonBody(c);
  const command = String(payload.command ?? "").trim();
  if (!command) return fail(c, "command is required");
  c.set("auditDetail", command.slice(0, 500));
  try {
    return ok(c, await providerFor(comp).execInDesktopResult(comp.slug, command));
  } catch (err: any) {
    return fail(c, String(err?.message ?? err), 500);
  }
});
